OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,1142 @@
+};
+
+class OpenSim::ColumnHasLabel : public std::runtime_error {
+public:
+  ColumnHasLabel(const std::string& expl) : runtime_error(expl) {}
+};
+
+class OpenSim::ColumnHasNoLabel : public std::runtime_error {
+public:
+  ColumnHasNoLabel(const std::string& expl) : runtime_error(expl) {}
+};
+
+class OpenSim::ZeroElements : public std::runtime_error {
+public:
+  ZeroElements(const std::string& expl) : runtime_error(expl) {}
+};
+
+class OpenSim::InvalidEntry : public std::runtime_error {
+public:
+  InvalidEntry(const std::string& expl) : runtime_error(expl) {}
+};
+
+class OpenSim::MetaDataKeyExists : public std::runtime_error {
+public:
+  MetaDataKeyExists(const std::string& expl) : runtime_error(expl) {}
+};
+
+class OpenSim::MetaDataKeyDoesNotExist : public std::runtime_error {
+public:
+  MetaDataKeyDoesNotExist(const std::string& expl) : runtime_error(expl) {}
+};
+
+class OpenSim::MetaDataTypeMismatch : public std::runtime_error {
+public:
+  MetaDataTypeMismatch(const std::string& expl) : runtime_error(expl) {}
+};
+
+
+/**----------------------------------------------------------------------------*
+* \brief DataTable_ is a in-memory storage container for data(in the form of a *
+* matrix with column names) with support for holding metatdata.                *
+*                                                                              *
+* <ul> <li>Underlying matrix will have entries of configurable type            *
+* ET(template param).</li>  <li>Random-access(constant-time) to specific       *
+* entries, entire columns and entire rows using their index.</li>              *
+* <li>Average constant-time access to columns through column-labels.</li>      *
+* <li>Add rows and columns to existing DataTable_.</li> <li>Bind two           *
+* DataTable_(s) by row and by column. </li> <li>Set column labels for a subset *
+* of columns, update them, remove them etc.</li> <li>Construct DataTable_      *
+* emtpy OR with a given shape and default value OR using and iterator pair     *
+* producing one entry at a time.</li> <li>Heterogeneous metadata container.    *
+* Metadata in the form of key-value pairs where key is a std::string and       *
+* value is is of any type.</li> </ul>                                          *
+*                                                                              *
+* \tparam ET Type of the entries in the underlying matrix. Defaults to         *
+* SimTK::Real(alias for double).                                               *
+*-----------------------------------------------------------------------------*/
+template<typename ET = SimTK::Real> // Element datatype.
+class OpenSim::DataTable_ : public OpenSim::AbstractDataTable {
+public:
+  using size_t                     = std::size_t;
+  using string                     = std::string;
+  using MetaDataType               = std::unordered_map<string, 
+                                                        SimTK::AbstractValue*>;
```

Comment:
> Why type in MetaDataType and not just MetaData?

I agree that adding "Type" to the end doesn't add anything. We don't usually put in the names anything that can be determined from the type. Some coding conventions do include types in the name (most famously, Microsoft's [Hungarian notation](https://en.wikipedia.org/?title=Hungarian_notation)) but we explicitly rejected that idea in ours (see [here](https://github.com/simbody/simbody/blob/master/CONTRIBUTING.md#naming-conventions)).